datamodel/include/datamodel/math/Vector.hpp

```diff
@@ -262,10 +259,59 @@ template <>
 void to_binary(BinaryBuffer &b, const rcbe::math::Vector3f &v);
 }
 
-rcbe::core::EngineScalar operator*(const rcbe::math::Vector3d &lhs, const rcbe::math::Vector3d &rhs);
-rcbe::math::Vector3d operator+(const rcbe::math::Vector3d &lhs, const rcbe::math::Vector3d &rhs);
-rcbe::math::Vector3d operator-(const rcbe::math::Vector3d &lhs, const rcbe::math::Vector3d &rhs);
-rcbe::math::Vector3d operator*(const rcbe::core::EngineScalar &lhs, const rcbe::math::Vector3d &rhs);
-rcbe::math::Vector3d operator*(const rcbe::math::Vector3d &lhs, const rcbe::core::EngineScalar &rhs);
+template <typename V, size_t D>
+typename rcbe::math::Vector<V, D>::ValueType operator*(const rcbe::math::Vector<V, D> &lhs, const rcbe::math::Vector<V, D> &rhs) {
+    typename rcbe::math::Vector<V, D>::ValueType ret {};
+
+    for (size_t i = 0; i < D; ++i) {
+        ret += lhs.at(i) * rhs.at(i);
+    }
+
+    return ret;
+}
+
+template <typename V, size_t D>
+rcbe::math::Vector<V, D> operator+(const rcbe::math::Vector<V, D> &lhs, const rcbe::math::Vector<V, D> &rhs) {
+    rcbe::math::Vector<V, D> ret{};
+
+    for (size_t i = 0; i < D; ++i) {
+        ret[i] = lhs[i] + rhs[i];
+    }
+
+    return ret;
+}
+
+template <typename V, size_t D>
+rcbe::math::Vector3d operator-(const rcbe::math::Vector<V, D> &lhs, const rcbe::math::Vector<V, D> &rhs) {
+    rcbe::math::Vector<V, D> ret{};
+
+    for (size_t i = 0; i < D; ++i) {
+        ret[i] = lhs[i] - rhs[i];
+    }
+
+    return ret;
+}
+
+template <rdmn::core::NumberValue NV, typename V, size_t D>
```

Comment:
Order of arguments in the template clause is not essential, although I would align them with the op's signature.

Reply:
Fair enough